File: working_hours/application/file_service.py
"""Raw data file management use cases."""
from __future__ import annotations

from werkzeug.utils import secure_filename

from ..domain.parser import validate_raw_content
from ..infrastructure.data import _raw_dir


def list_raw_files() -> list[dict]:
    d = _raw_dir()
    if not d.exists():
        return []
    return [
        {"name": f.name, "size": f.stat().st_size, "modified": f.stat().st_mtime}
        for f in sorted(d.glob("*.txt"))
    ]


def save_raw_file(name: str, content: str) -> tuple[bool, str]:
    """Validate and persist a raw data file. Returns (ok, error_message)."""
    name = secure_filename(name)
    if not name:
        return False, "No filename."
    if not name.lower().endswith(".txt"):
        return False, "Only .txt files are accepted."
    ok, error = validate_raw_content(content)
    if not ok:
        return False, error
    d = _raw_dir()
    d.mkdir(parents=True, exist_ok=True)
    (d / name).write_text(content, encoding="utf-8")
    return True, ""


def delete_raw_file(filename: str) -> tuple[bool, str]:
    """Delete a raw data file. Returns (ok, error_message)."""
    name = secure_filename(filename)
    if not name.lower().endswith(".txt"):
        return False, "Invalid file."
    p = _raw_dir() / name
    if not p.exists():
        return False, "File not found."
    p.unlink()
    return True, ""

File: packages/server/src/application/fileService.ts
import { existsSync, mkdirSync, readdirSync, statSync, unlinkSync, writeFileSync } from 'fs'
import { join } from 'path'
import { validateRawContent } from '../domain/parser.js'
import { rawDir } from '../infrastructure/data.js'

function secureFilename(name: string): string {
  return name
    .split(/[\\/]/).pop()!
    .replace(/[^a-zA-Z0-9._-]/g, '_')
    .replace(/^[._]+/, '')
}

export function listRawFiles(): object[] {
  const d = rawDir()
  if (!existsSync(d)) return []
  return readdirSync(d)
    .filter(n => n.endsWith('.txt'))
    .sort()
    .map(n => {
      const st = statSync(join(d, n))
      return { name: n, size: st.size, modified: st.mtimeMs / 1000 }
    })
}

export function saveRawFile(name: string, content: string): { ok: boolean; error: string } {
  const safe = secureFilename(name)
  if (!safe) return { ok: false, error: 'No filename.' }
  if (!safe.toLowerCase().endsWith('.txt')) return { ok: false, error: 'Only .txt files are accepted.' }
  const { ok, error } = validateRawContent(content)
  if (!ok) return { ok: false, error }
  const d = rawDir()
  mkdirSync(d, { recursive: true })
  writeFileSync(join(d, safe), content, 'utf-8')
  return { ok: true, error: '' }
}

export function deleteRawFile(filename: string): { ok: boolean; error: string } {
  const safe = secureFilename(filename)
  if (!safe.toLowerCase().endsWith('.txt')) return { ok: false, error: 'Invalid file.' }
  const p = join(rawDir(), safe)
  if (!existsSync(p)) return { ok: false, error: 'File not found.' }
  unlinkSync(p)
  return { ok: true, error: '' }
}
